cpp/foxglove/include/foxglove/channel.hpp

```diff
@@ -35,6 +35,55 @@ struct Schema {
   size_t data_len = 0;
 };
 
+/// @brief A description of a channel. This will be constructed by the SDK and passed to an
+/// implementation of a `SinkChannelFilterFn`.
+class ChannelDescriptor {
+  std::string topic_;
+  std::string message_encoding_;
+  std::optional<std::string> schema_name_;
+  std::optional<std::string> schema_encoding_;
+  std::optional<std::map<std::string, std::string>> metadata_;
+
+public:
+  // @cond foxglove_internal
+  ChannelDescriptor(
+    std::string topic, std::string message_encoding, std::optional<std::string> schema_name,
+    std::optional<std::string> schema_encoding,
+    std::optional<std::map<std::string, std::string>> metadata = std::nullopt
+  );
+  // @endcond
+
+  /// @brief Get the topic of the channel.
+  [[nodiscard]] const std::string& topic() const noexcept;
+
+  /// @brief Get the message encoding of the channel.
+  [[nodiscard]] const std::string& message_encoding() const noexcept;
+
+  /// @brief Get the metadata for the channel.
+  [[nodiscard]] const std::optional<std::map<std::string, std::string>>& metadata() const noexcept;
+
+  /// @brief Get the schema name of the channel.
+  [[nodiscard]] const std::optional<std::string>& schema_name() const noexcept;
+  /// @brief Get the schema encoding of the channel.
+  [[nodiscard]] const std::optional<std::string>& schema_encoding() const noexcept;
+
+  // @cond foxglove_internal
+  ChannelDescriptor(ChannelDescriptor&& other) noexcept = default;
+  ChannelDescriptor& operator=(ChannelDescriptor&& other) noexcept = default;
+  ~ChannelDescriptor() = default;
+  // Disable copy operations
+  ChannelDescriptor(const ChannelDescriptor&) = delete;
```

Comment:
Curious as to why copying these around is forbidden. They don't own any resources and boil down to a handful of strings. It's probably not the most efficient thing in the world to copy them if you don't have a good reason, but I don't know if I love globally forcing move constructors.